### PR TITLE
hw5 PoA

### PR DESCRIPTION
**Pull request by jashonnew**
I am planning to follow the unsloth training example that Chris provided, but on the smaller model. I want to just train the data on the starwars data and compare answers to a simple question before and after the training. I will use cross entropy as my metric for training differential. The is going to be a 'finetuning' training.  I could use a separate model to try and do some sort of knowledge distillation, but I think for an initial LLM training it makes the most sense for me to try and recreate what has been done and get a grasp on the fundamentals. (please do not grade PR until I have commented that I am finished with ASSIGNMENT 5)

### CODE REVIEW

**Comment by jashonnew** [comment]
I added my Kaggle notebook in which I successfully adapted Chris' code to finetune the llama 8b on the star wars dataset. I regret not checking the models answer to my question initially before finetuning, this would have given me a good baseline for how well the model performed at the task before I adjusted it. 

I also added a run.py file that includes all of the code from the ipynb but without any of the markdown lines so that you can easily analyze my process. In total training took 2.8 hours, and the process of adapting Chris' work and understanding what was going on took me somewhere in the arena of 8+ hours. 

I had chatgpt write tests, but since I don't have access to a GPU locally I am not sure these tests mean much for my environment. tests handled:

- test generation output
- test model saving and loading
- test training process
- test model forward pass
- test data preprocessing
- test tokenization
- test dataset loading

I am not sure that I could articulate what the loss means in this context as we finetune. Over the 2300 lines of data from the starwars text it looks like loss went from 11 to somewhere around 5, but I don't fully understand the interpretation on that.

I did use ruff to format my code. 5 files changed. I would also like to take time to attempt the extra credit for this week and refactor previous code, as I am somewhat behind currently I am just trying to get my feet under me before making sure I have the right shoes on. If that makes sense, LOL.

**Comment by msharp9** [comment]
Great work. I'm going to close this for now, but if you'd like to go back and do the extra credit, just reopen the PR.
